File: flutter_readium/android/src/main/kotlin/dk/nota/flutter_readium/FlutterPdfPreferences.kt
package dk.nota.flutter_readium

import org.readium.adapter.pdfium.navigator.PdfiumPreferences
import org.readium.r2.navigator.preferences.Axis
import org.readium.r2.navigator.preferences.Fit
import org.readium.r2.navigator.preferences.ReadingProgression
import org.readium.r2.shared.ExperimentalReadiumApi

// PdfiumPreferences in kotlin-toolkit has no paginated mode — Pdfium is
// always continuous-scroll. The Dart-side `PDFLayout.paginated` is therefore
// mapped to `scrollAxis = HORIZONTAL`, which gives the closest equivalent
// (single-page-wide viewport, one page per swipe). `scrollHorizontal` maps to
// the same axis. `scrollVertical` maps to `Axis.VERTICAL`.
//
// Dart-side `PDFFit.page` maps to Pdfium `Fit.CONTAIN`. `PDFFit.auto` has no
// Pdfium equivalent and is ignored on Android.
@OptIn(ExperimentalReadiumApi::class)
data class FlutterPdfPreferences(
    val readingProgression: ReadingProgression? = null,
    val scrollAxis: Axis? = null,
    val fit: Fit? = null,
    val pageSpacing: Double? = null,
) {
    fun toPdfiumPreferences(): PdfiumPreferences =
        PdfiumPreferences(
            fit = fit,
            pageSpacing = pageSpacing,
            readingProgression = readingProgression,
            scrollAxis = scrollAxis,
        )

    companion object {
        fun fromMap(map: Map<String, Any>): FlutterPdfPreferences {
            val rpStr = map["readingProgression"] as? String
            val readingProgression =
                rpStr?.let {
                    when (it) {
                        "ltr" -> ReadingProgression.LTR
                        "rtl" -> ReadingProgression.RTL
                        else -> null
                    }
                }
            val layoutStr = map["layout"] as? String
            val scrollAxis =
                layoutStr?.let {
                    when (it) {
                        "paginated", "scrollHorizontal" -> Axis.HORIZONTAL
                        "scrollVertical" -> Axis.VERTICAL
                        else -> null
                    }
                }
            val fitStr = map["fit"] as? String
            val fit =
                fitStr?.let {
                    when (it) {
                        "page", "contain" -> Fit.CONTAIN

                        "width" -> Fit.WIDTH

                        // Kotlin toolkit Pdfium supports only CONTAIN and WIDTH.
                        "auto" -> null

                        else -> null
                    }
                }
            val pageSpacing = (map["pageSpacing"] as? Number)?.toDouble()
            return FlutterPdfPreferences(
                readingProgression = readingProgression,
                scrollAxis = scrollAxis,
                fit = fit,
                pageSpacing = pageSpacing,
            )
        }
    }
}

File: flutter_readium/ios/flutter_readium/Sources/flutter_readium/FlutterPDFPreferences.swift
import Foundation
import ReadiumNavigator
import ReadiumShared

/// Maps the Dart PDFPreferences JSON map to ReadiumNavigator's
/// PDFNavigatorViewController.Preferences for runtime updates.
struct FlutterPDFPreferences {
  var readium: PDFNavigatorViewController.Preferences = .init()

  init() {}

  init(fromMap map: [String: Any]) {
    var prefs = PDFNavigatorViewController.Preferences()
    // Unified PDFLayout enum from Dart, translated into PDFKit's `scroll` +
    // `scrollAxis` pair. `paginated` uses scroll=false (snap-to-page);
    // the two `scroll*` cases use scroll=true with the matching axis.
    if let layoutString = map["layout"] as? String {
      switch layoutString {
      case "paginated":
        prefs.scroll = false
      case "scrollVertical":
        prefs.scroll = true
        prefs.scrollAxis = .vertical
      case "scrollHorizontal":
        prefs.scroll = true
        prefs.scrollAxis = .horizontal
      default:
        break
      }
    }
    if let rpString = map["readingProgression"] as? String,
       let rp = ReadiumNavigator.ReadingProgression(rawValue: rpString) {
      prefs.readingProgression = rp
    }
    if let fitString = map["fit"] as? String,
       let fit = ReadiumNavigator.Fit(rawValue: fitString) {
      prefs.fit = fit
    }
    if let pageSpacing = map["pageSpacing"] as? NSNumber {
      prefs.pageSpacing = pageSpacing.doubleValue
    }
    if let offsetFirstPage = map["offsetFirstPage"] as? Bool {
      prefs.offsetFirstPage = offsetFirstPage
    }
    if let spreadString = map["spread"] as? String,
       let spread = ReadiumNavigator.Spread(rawValue: spreadString) {
      prefs.spread = spread
    }
    if let visibleScrollbar = map["visibleScrollbar"] as? Bool {
      prefs.visibleScrollbar = visibleScrollbar
    }
    readium = prefs
  }
}
